pkcs11/src/new/functions/object_management.rs

```diff
@@ -110,58 +116,60 @@ impl Pkcs11 {
             })
             .collect();
 
-        Rv::from(unsafe {
-            ((*self.function_list).C_GetAttributeValue.unwrap())(
+        unsafe {
+            Rv::from(get_pkcs11!(self, C_GetAttributeValue)(
                 session.handle(),
                 object.handle(),
                 test_template.as_mut_ptr(),
                 test_template.len().try_into()?,
-            )
-        })
-        .to_result()
-        .or_else(|e| {
-            for attribute in &test_template {
-                if attribute.ulValueLen == CK_UNAVAILABLE_INFORMATION {
-                    error!("Attribute {} is unavailable.", attribute.type_);
+            ))
+            .into_result()
+            .map_err(|e| {
+                for attribute in &test_template {
+                    if attribute.ulValueLen == CK_UNAVAILABLE_INFORMATION {
+                        error!("Attribute {} is unavailable.", attribute.type_);
+                    }
                 }
-            }
-            Err(e)
-        })?;
+                e
+            })?;
+        }
 
         // Check that the length of the attribute is the same one as in the originial iterator
         template
             .iter()
             .cloned()
             .zip(test_template.into_iter())
-            .map(|(given, expected)| {
-                if given.ulValueLen < expected.ulValueLen {
-                    error!(
-                        "Attribute of type {} has a buffer too small. {} expected, {} given.",
-                        given.type_, expected.ulValueLen, given.ulValueLen
-                    );
-                    Err(RvError::BufferTooSmall.into())
-                } else if given.ulValueLen > expected.ulValueLen {
-                    error!(
-                        "Attribute of type {} has a buffer too big. {} expected, {} given.",
-                        given.type_, expected.ulValueLen, given.ulValueLen
-                    );
-                    Err(Error::BufferTooBig)
-                } else {
-                    Ok(())
-                }
-            })
+            .map(
+                |(given, expected)| match given.ulValueLen.cmp(&expected.ulValueLen) {
+                    Ordering::Less => {
+                        error!(
+                            "Attribute of type {} has a buffer too small. {} expected, {} given.",
+                            given.type_, expected.ulValueLen, given.ulValueLen
+                        );
+                        Err(RvError::BufferTooSmall.into())
+                    }
+                    Ordering::Greater => {
+                        error!(
+                            "Attribute of type {} has a buffer too big. {} expected, {} given.",
+                            given.type_, expected.ulValueLen, given.ulValueLen
+                        );
+                        Err(Error::BufferTooBig)
+                    }
```

Comment:
The spec for this function seems to say:
> 4. Otherwise, if the length specified in ulValueLen is large enough to hold the value of the specified attribute for the object, then that attribute is copied into the buffer located at pValue, and the ulValueLen field is modified to hold the exact length of the attribute.

Is that relevant here?

Reply:
Yes it is the case covered by this arm. The reason I did this was to force callers of this function to give buffers with size matching exactly the the attributes that they want. Otherwise we would have to "resize" the slices that the Rust callers give, not sure if that is possible.

Reply:
Are callers able to always give the correct size, without using the function with `null()` for pointers?

Reply:
That's why I added the `get_attribute_info` method below to just get the size information for a vector of `AttributeType` 👌

Reply:
Kewl, that would go well in the docs of this method, then